sympy/matrices/matrices.py

```diff
@@ -2517,10 +2517,14 @@ def _jblock_exponential(b):
 
         blocks = list(map(_jblock_exponential, cells))
         from sympy.matrices import diag
+        from sympy import re
         eJ = diag(*blocks)
         # n = self.rows
         ret = P * eJ * P.inv()
-        return type(self)(ret)
+        if all(value.is_real for value in self.values()):
+            return re(ret)
```

Comment:
This should be `type(self)(re(ret))`. Some matrices are mutable, some are immutable. The reason for `type(self)` is that exp(M) should have the same type as M.   As written, this code will return an immutable matrix even if self was mutable.